src/v/raft/coordinated_recovery_throttle.cc

```diff
@@ -21,46 +21,75 @@
 #include <boost/range/irange.hpp>
 
 #include <chrono>
+#include <cstdio>
+#include <iterator>
+#include <limits>
 
 namespace raft {
 
 coordinated_recovery_throttle::token_bucket::token_bucket(
   size_t initial_capacity)
-  : _sem(initial_capacity, "recovery_throttle")
-  , _last_reset_capacity(initial_capacity) {}
+  : _available_units(initial_capacity)
+  , _last_reset_capacity(initial_capacity) {
+    _cv.signal();
+}
 
 ss::future<> coordinated_recovery_throttle::token_bucket::throttle(
   size_t size, ss::abort_source& as) {
     _waiting_bytes += size;
     auto decrement = ss::defer([this, size] { _waiting_bytes -= size; });
-    co_await _sem.wait(as, size);
-    _admitted_bytes_since_last_reset += size;
+    auto lock = co_await _mutex.get_units(as);
+    while (true) {
+        co_await _cv.wait(as);
```

Comment:
Wondering why this is not _cv.wait(predicate). Without somehting like that a first request would be blocked until the next tick, right?

Reply:
Please ignore my previous comment here, it is not accurate.

If a condition variable (at least in Seastar) is signalled in absence of waiters this signal is not lost but rather memorised, so the first request to wait will immediately resolve. So even if requests come sparse within a tick they'll still all resolve immediately as long as there's allowance left.

Reply:
So the performance question is down to rather subtle things:
- cv.wait() with predicate will first check predicate without waiting for cv. It means that for a stuck oversized allocation it'll check it twice.
- OTOH it uses `ss::do_until` which is faster than `co_await`.